integration-tests/sample_test.go

```diff
@@ -58,3 +62,156 @@ func TestAuth(t *testing.T) {
 		t.Fatal("UserID  mismatch")
 	}
 }
+
+func TestValidateSignature(t *testing.T) {
+	if options.proxyHost == "" || options.proxyPort == "" || options.sampleName == "" {
+		t.Skip("sample-name, proxy-host or proxy-port are missing => skip TestValidateSignature")
+	}
+
+	// All samples generate a new state and redirect us to an /authorize URL, if we're not logged in.
+	httpClient := &http.Client{
+		CheckRedirect: func(req *http.Request, via []*http.Request) error {
+			return http.ErrUseLastResponse
+		},
+	}
+
+	client := newSampleClient(options.sampleURL, httpClient)
+	restartSample(client)
+
+	m := newModifier(modifySignatureHandler)
+	m.start()
+
+	defer m.stop()
+
+	payload := struct {
+		ModifyURL string `json:"modifyUrl"`
+	}{
+		ModifyURL: "http://127.0.0.1:" + fmt.Sprint(m.getPort()),
+	}
+
+	prreq, err := newRequest(fmt.Sprintf("http://%v:%v/session", options.proxyHost, options.proxyPort), "POST", payload)
+	if err != nil {
+		t.Fatalf("Error making request to the proxy: %v", err)
+	}
+
+	_, _, err = getResponse(prreq, httpClient)
+	if err != nil {
+		t.Fatalf("Error when receiving the proxy response: %v", err)
+	}
+
+	name := uuid.New().String()
+	userID := fmt.Sprintf("%v@example.com", name)
+	deviceName := "The device of " + name
+	pin := randPIN()
+
+	authorizeRequestURL, err := client.authorize()
+	if err != nil {
+		t.Fatalf("Error making initial request to the home page: %v", err)
+	}
+
+	_, err = url.ParseRequestURI(string(authorizeRequestURL))
+	if err != nil {
+		t.Fatalf("Home page didn't respond with authorize url, error: %v, actual response: %v", err.Error(), string(authorizeRequestURL))
+	}
+
+	identity, err := register(httpClient, userID, deviceName, pin, string(authorizeRequestURL))
+	if err != nil {
+		t.Fatalf("Error registering: %v", err)
+	}
+
+	accessResponse, err := authenticate(httpClient, identity, userID, pin, string(authorizeRequestURL))
+	if err != nil {
+		t.Fatalf("Error authenticating: %v", err)
+	}
+
+	// Contains an URL, from which we can fetch the user info.
+	err = client.login(accessResponse.RedirectURL)
+	if err == nil {
+		t.Fatal("Expected error when the signature is invalid")
+	}
+
+	if !errors.Is(err, errInvalidLogin) {
+		t.Fatal("Expected invalid login error")
+	}
+
+	if !strings.Contains(err.Error(), "Invalid ID token") &&
+		!strings.Contains(err.Error(), "Signature validation failed.") &&
+		!strings.Contains(err.Error(), "invalid login") {
+		t.Fatalf("Expected Invalid ID token error message")
+	}
```

Comment:
That's fragile. There need to be different errors for different cases. I also don't understand why  `errInvalidLogin` is wrapped.